apps/api/src/donations/events/stripe-payment.service.ts

```diff
@@ -110,6 +114,157 @@ export class StripePaymentService {
+    }
+
+    if (!metadata.personId) {
+      throw new BadRequestException('Subscription metadata does not contain target personId. Subscription is: ' + subscription)
+    }
+
+    const recurringDonation = await this.recurringDonationService.findSubscriptionByExtId(subscription.id)
+    if (!recurringDonation) {
+      this.handleSubscriptionCreated(event)
+      return
+    }
+
+    Logger.debug('Updating recurring donation by id ' + recurringDonation.id)
+
+    this.recurringDonationService.updateStatus(recurringDonation.id, string2RecurringDonationStatus(subscription.status))
+  }
+
+  @StripeWebhookHandler('customer.subscription.deleted')
+  async handleSubscriptionDeleted(event: Stripe.Event) {
+    const subscription: Stripe.Subscription = event.data.object as Stripe.Subscription
+    Logger.log('[ handleSubscriptionDeleted ]', subscription)
+
+    const metadata: DonationMetadata = subscription.metadata as DonationMetadata
+    if (!metadata.campaignId) {
+      throw new BadRequestException('Subscription metadata does not contain target campaignId. Subscription is: ' + subscription)
+    }
+
+    const recurringDonation = await this.recurringDonationService.findSubscriptionByExtId(subscription.id)
+    if (!recurringDonation) {
+      Logger.debug('Received a notification about unknown subscription: ' + subscription.id)
+      return
+    }
+
+    Logger.debug('Deleting recurring donation by id ' + recurringDonation.id)
+
+    this.recurringDonationService.updateStatus(recurringDonation.id, string2RecurringDonationStatus(subscription.status))
+  }
+
+  @StripeWebhookHandler('invoice.paid')
+  async handleInvoicePaid(event: Stripe.Event) {
+    const invoice: Stripe.Invoice = event.data.object as Stripe.Invoice
+    Logger.log(
+      '[ handleInvoicePaid ]',
+      invoice,
+    )
+
+    let metadata: DonationMetadata = { campaignId: null, personId: null }
+
+    invoice.lines.data.forEach((line: Stripe.InvoiceLineItem) => {
+      if (line.type === 'subscription') {
+        metadata = line.metadata as DonationMetadata
+      }
+    })
+
+    if (!metadata.campaignId) {
+      throw new BadRequestException(
+        'Invoice intent metadata does not contain target campaignId. Probably wrong session initiation. Invoice id is:  ' +
+          invoice.id,
+      )
+    }
+
+    const campaign = await this.campaignService.getCampaignById(metadata.campaignId)
+
+    if (campaign.currency !== invoice.currency.toUpperCase()) {
+      throw new BadRequestException(
+        `Invoice in different currency is not allowed. Campaign currency ${
+          campaign.currency
+        } <> donation currency ${invoice.currency.toUpperCase()}`,
+      )
+    }
+
+    const billingData = getInvoiceData(invoice)
+
+    this.campaignService.donateToCampaign(campaign, billingData).then(() => {
+      this.checkForCompletedCampaign(metadata.campaignId as string)
+    })
```

Comment:
I don't get why we use `then()` syntnax here instead of `await this.campaignService.donateToCampaign`

This breaks the synchronous flow in my opinion.